### ISSUE TITLE
Rule Change: no-eval flags eval in TypeScript type positions

### ISSUE THREAD

**Issue by sethamus** [eslint/eslint#20052]
### What rule do you want to change?

no-eval

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint no-eval: "error" */

let myEvalFunction: typeof eval;
```

### What does the rule currently do for this code?

Reports an error: “eval can be harmful.” The identifier is flagged even though it appears only in a type position, not at runtime.

### What will the rule do after it's changed?

Do not report when eval appears only in type positions (e.g., within typeof, type aliases, interfaces, etc.).

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
I think this is an edge case that wouldn't justify added complexity. If reporting references to `eval` in types isn't desirable, user can add eslint-disable comments where they are needed.

@eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
What complexity are you referring to? Couldn't we just do `TSTypeAnnotation Identifier[name=eval]` and filter those results out?

**Comment by mdjermanovic** [comment]
> Couldn't we just do `TSTypeAnnotation Identifier[name=eval]` and filter those results out?

These references can appear not only within `TSTypeAnnotation` nodes. For example:

```ts
type foo = typeof eval;
```

However, maybe we could just check if Identifier's parent is `TSTypeQuery`? If the change that covers all cases would be as simple as that, I'd be in favor.

**Comment by fasttime** [comment]
I was curious, so I searched for `typeof eval` on GitHub and found several occurrences, mostly in forks of DefinitelyTyped: https://github.com/search?q=repo%3ADefinitelyTyped%2FDefinitelyTyped+%2Ftypeof+eval%5Cb%2F&type=code.

It turns out this pattern isn't as rare as I expected, with most usages appearing in property definitions (like `eval: typeof eval;`) and in type parameters (`expectType<typeof eval>`). Probably it should be sufficient to check if the `Identifier`'s parent is `TSTypeQuery` as suggested by @mdjermanovic to detect these cases. I'd be mildly in favor of this change.

**Comment by JoshuaKGoldberg** [comment]
🤔 One could make the argument that asking for the type of `eval` itself should also be bad: why would you need that? If code is reusing `eval`'s type, what purpose would that serve other than to effectively use `eval`?

I'm mildly against this change, but could easily be swayed with some real-world examples that show behavior we wouldn't want to lint against.

**Comment by nzakas** [comment]
> One could make the argument that asking for the type of `eval` itself should also be bad:why would you need that? If code is reusing `eval`'s type, what purpose would that serve other than to effectively use `eval`?

I find that to be a compelling argument. :+1:

**Comment by sethamus** [comment]
I still think `no-eval` should only flag runtime `eval` usage.

> Probably it should be sufficient to check if the Identifier's parent is TSTypeQuery as suggested by @mdjermanovic to detect these cases.

Agreed — checking `Identifier` with parent `TSTypeQuery` should be sufficient.
